design/service-apis-implementation.md

```diff
@@ -0,0 +1,229 @@
+- Cluster operator: The cluster operator (ops) is responsible for administration of entire clusters. They manage policies, network access, application permissions.
+- Application developer: The application developer (dev) is responsible for defining their application configuration (e.g. timeouts, request matching/filter) and Service composition (e.g. path routing to backends).
+
+The cluster operator and application developer are basically the same as Contour's Cluster Administrator and Application Developer personas, which will be important for this design.
+
+In terms of the APIs themsleves, the Service APIs have 3 primary API resources (taken from the service-apis docs site):
+
+- GatewayClass defines a set of gateways with a common configuration and behavior.
+- Gateway requests a point where traffic can be translated to Services within the cluster.
+- Routes describe how traffic coming via the Gateway maps to the Services.
+
+In Contour, we've previously solved a lot of the same problems with HTTPProxy (and IngressRoute before it).
+That functionality can be described in the Service APIs by HTTPRoutes, HTTPSRoutes and TLSRoutes, as they describe Layer 7 ingress, as Contour does today.
+
+Other types of routes include TCPRoutes and UDPRoutes, which are intended for Layer 4 load balancers. Implementations may also define their own Route objects, at the cost of interoperability.
+
+In terms of its stated goals, Contour is aiming at being an ingress controller - that is, a Layer 7 proxy with some api gateway functions.
+Currently, Contour provides "TCP Proxying" that allows the forwarding of TLS streams based on the SNI information, which is precisely what the Service APIs TLSRoute object is for.
+If Project Contour (the organisation) does add support for TCP and UDP forwarding, it will not be in the `projectcontour/contour` repo, but will be a separate repo.
+
+## Goals
+- Define a data model and implementation for Contour's service-apis support.
+- Layer 7 support only, which means HTTPRoutes, HTTPSRoutes, and TLSRoutes only.
+
+## Non Goals
+- No TCPRoute or UDPRoute support, that is, no support for arbitrary TCP/UDP forwarding.
+
+
+## High-Level Design
+
+### Service APIs implementation surface area
+
+Contour's support for the Service APIs will include support for the HTTPRoute and TLSRoute objects only, not any other type of Route - this includes no TCPRoute or UDPRoute support.
+Contour is a layer 7 ingress controller, and the layer 4 load balancing implies by TCPRoute and UDPRoute is out of scope for this tool (that is, `projectcontour/contour`).
+Project Contour (the organisation) may investigate a Service-APIs based layer 4 solution in the future, but that effort will not be in `projectcontour/contour`.
+
+### Contour Gateway model
+
+Contour considers a GatewayClass and associated set of Gateways to describe a single Envoy deployment, deployed and managed by something outside of itself.
```

Comment:
can you remind me of the scenario where a Contour user would have more than one `Gateway`?

Reply:
Looks like we have significant conflicts between the operator proposal and this one: https://github.com/projectcontour/contour-operator/pull/171/files#diff-1cbda5ba0844ea1d442d220ece0cd2f53066104e33808dffb0b7d8df6c2ff5d8R106

Reply:
cc @danehans

Reply:
https://github.com/projectcontour/contour-operator/pull/171/files#diff-1cbda5ba0844ea1d442d220ece0cd2f53066104e33808dffb0b7d8df6c2ff5d8R114

Reply:
As discussed in our maintainer meeting, Contour will move to being configured to check a *specific*, named by namespace and name Gateway only. This is because of this line. I'll update this to make it clearer.